src/EmbedIO/Security/IPBanningModule.cs

```diff
@@ -0,0 +1,301 @@
+                Blacklist.ContainsKey(ClientAddress))
+                return;
+
+            foreach (var regex in FailRegex.Values)
+            {
+                try
+                {
+                    if (!regex.IsMatch(logEvent.Message)) continue;
+
+                    // Add to list
+                    AddAccessAttempt(ClientAddress);
+                    UpdateBlackList();
+                    break;
+                }
+                catch (RegexMatchTimeoutException ex)
+                {
+                    $"Timeout trying to match '{ex.Input}' with pattern '{ex.Pattern}'.".Error(nameof(IPBanningModule));
+                }
+            }
+        }
+        
+        /// <inheritdoc />
+        public void Dispose() =>
+            Dispose(true);
+
+        internal void AddRules(IEnumerable<string> patterns)
+        {
+            foreach (var pattern in patterns)
+                AddRule(pattern);
+        }
+
+        internal void AddRule(string pattern)
+        {
+            try
+            {
+                FailRegex.TryAdd(pattern, new Regex(pattern, RegexOptions.Compiled | RegexOptions.CultureInvariant, TimeSpan.FromMilliseconds(500)));
+            }
+            catch (Exception ex)
+            {
+                ex.Log(nameof(IPBanningModule), $"Invalid regex - '{pattern}'.");
+            }
+        }
+
+        internal void AddToWhitelist(IEnumerable<string> whitelist) =>
+            AddToWhitelistAsync(whitelist).GetAwaiter().GetResult();
+
+        internal async Task AddToWhitelistAsync(IEnumerable<string> whitelist)
+        {
+            if (whitelist?.Any() != true)
+                return;
+
+            foreach (var address in whitelist)
+            {
+                var addressees = await IPParser.Parse(address).ConfigureAwait(false);
+                _whitelist.AddRange(addressees.Where(x => !_whitelist.Contains(x)));
+            }
+        }
+        
+        /// <inheritdoc />
+        protected override Task OnRequestAsync(IHttpContext context)
+        {
+            ClientAddress = context.Request.RemoteEndPoint.Address;
+            if (!Blacklist.ContainsKey(ClientAddress)) 
+                return Task.CompletedTask;
+
+            context.SetHandled();
+            throw HttpException.Forbidden();
+        }
+        
+        /// <summary>
+        /// Releases unmanaged and - optionally - managed resources.
+        /// </summary>
+        /// <param name="disposing"><c>true</c> to release both managed and unmanaged resources; <c>false</c> to release only unmanaged resources.</param>
+        protected virtual void Dispose(bool disposing)
+        {
+            if (_disposedValue) return;
+            if (disposing)
+            {
+                _whitelist.Clear();
+            }
+
+            _disposedValue = true;
+        }
+
+        private static void AddAccessAttempt(IPAddress address)
+        {
+            if (AccessAttempts.ContainsKey(address))
```

Comment:
You are throwing thread-safety away here, by accessing `AccessAttempts` in a non-atomic way.
The [`ConcurrentDictionary<TKey,TValue>.GetOrAdd`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd) method exists for precisely this kind of situation:

```C#
	AccessAttempts.GetOrAdd(address, _ => new ConcurrentBag<long>()).Add(DateTime.Now.Ticks);
```

If `address` is already a key in the dictionary, `GetOrAdd` will return its corresponding value; otherwise, it will call the lambda to get a value, add the key / value pair to the dictionary, and return the value.

Reply:
Thanks, I already changed this to use `GetOrAdd`